File: kmp-mobile/ios/PazChurch/Features/Academy/AcademyViewModel.swift
import Observation
import Shared
import SwiftUI

@MainActor
@Observable
class AcademyViewModel {
    var tracks: [CourseTrack] = []
    var isLoading = true
    var error: String?
    var resumeCourse: Course?
    var latestStudy: LifeGroupStudy?

    private let academyRepository: AcademyRepository
    private let lifeGroupStudyRepository: LifeGroupStudyRepository

    init(academyRepository: AcademyRepository, lifeGroupStudyRepository: LifeGroupStudyRepository) {
        self.academyRepository = academyRepository
        self.lifeGroupStudyRepository = lifeGroupStudyRepository
    }

    func load(isAuthenticated: Bool) async {
        isLoading = true
        error = nil
        do {
            let content = try await academyRepository.getAcademyContent()
            tracks = content.tracks
            // resumeCourse populated from last-watched logic when backend supports it
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false

        // Estudo do Life requires life-group membership server-side (canView) — a plain
        // 403 here (visitor, or a member not yet in a life group) shouldn't surface as a
        // page-level error, so this failure is silently ignored.
        guard isAuthenticated else {
            latestStudy = nil
            return
        }
        latestStudy = try? await lifeGroupStudyRepository.getStudies(page: 1, limit: 1).items.first
    }

    func onRetry(isAuthenticated: Bool) {
        Task { await load(isAuthenticated: isAuthenticated) }
    }

    func onCourseTapped(_ course: Course) {
        // navigate to video player — handled by parent NavigationStack
    }
}

File: kmp-mobile/android/src/main/kotlin/br/church/paz/android/ui/features/academy/AcademyViewModel.kt
package br.church.paz.android.ui.features.academy

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.church.paz.shared.domain.repository.AcademyRepository
import br.church.paz.shared.domain.repository.AuthRepository
import br.church.paz.shared.domain.repository.LifeGroupStudyRepository
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class AcademyViewModel(
    private val academyRepository: AcademyRepository,
    private val authRepository: AuthRepository,
    private val lifeGroupStudyRepository: LifeGroupStudyRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow(AcademyUiState())
    val uiState: StateFlow<AcademyUiState> = _uiState.asStateFlow()

    private val _effect = Channel<AcademyEffect>(Channel.BUFFERED)
    val effect = _effect.receiveAsFlow()

    init {
        load()
    }

    fun load() = fetch(showSkeleton = true)

    /** Pull-to-refresh entry point — re-invokes the same load path without the full-screen skeleton. */
    fun refresh() = fetch(showSkeleton = false)

    private fun fetch(showSkeleton: Boolean) {
        viewModelScope.launch {
            _uiState.update {
                it.copy(isLoading = showSkeleton, isRefreshing = !showSkeleton, error = null)
            }
            val user = runCatching { authRepository.currentUser() }.getOrNull()
            val isAuthenticated = user != null
            _uiState.update { it.copy(isAuthenticated = isAuthenticated) }
            runCatching { academyRepository.getAcademyContent() }
                .onSuccess { content ->
                    _uiState.update {
                        it.copy(isLoading = false, isRefreshing = false, tracks = content.tracks)
                    }
                }.onFailure { e ->
                    _uiState.update { it.copy(isLoading = false, isRefreshing = false, error = e.message) }
                }

            // Estudo do Life requires life-group membership server-side (canView) — a plain
            // 403 here (visitor, or a member not yet in a life group) shouldn't surface as a
            // page-level error, so this failure is silently ignored — same intent as iOS's `try?`.
            if (!isAuthenticated) {
                _uiState.update { it.copy(latestStudy = null) }
                return@launch
            }
            val latestStudy =
                runCatching { lifeGroupStudyRepository.getStudies(page = 1, limit = 1).items.firstOrNull() }
                    .onFailure { e -> Log.d("AcademyVM", "latestStudy fetch failed: ${e.message}") }
                    .getOrNull()
            _uiState.update { it.copy(latestStudy = latestStudy) }
        }
    }

    fun refreshAuthState() {
        viewModelScope.launch {
            val user = runCatching { authRepository.currentUser() }.getOrNull()
            _uiState.update { it.copy(isAuthenticated = user != null) }
        }
    }

    fun onVideoTapped(videoId: String) {
        viewModelScope.launch { _effect.send(AcademyEffect.NavigateToPlayer(videoId)) }
    }

    fun onCourseTapped(courseId: String) {
        viewModelScope.launch { _effect.send(AcademyEffect.NavigateToCourse(courseId)) }
    }

    fun onStudyTapped(studyId: String) {
        viewModelScope.launch { _effect.send(AcademyEffect.NavigateToStudy(studyId)) }
    }
}
